dags/pull_here.py

```diff
@@ -9,46 +9,119 @@
+     schedule_interval='30 16 * * * ' ,
+     catchup=False,
+     doc_md = doc_md,
+     tags=["HERE"]
+     )
+
+def pull_here():
+
+    @task
+    def send_request():
+        api_conn = BaseHook.get_connection('here_api_key')
+        access_token = get_access_token(api_conn.password, api_conn.extra_dejson['client_secret'], api_conn.extra_dejson['token_url'])
+        return access_token
+
+    @task
+    def get_request_id(access_token: str, **kwargs):
+        api_conn = BaseHook.get_connection('here_api_key')
+        ds = kwargs["ds"]
+        pull_date = ds_format(ds_add(ds, -1), "%Y-%m-%d", "%Y%m%d")
+        request_id = query_dates(access_token, pull_date, pull_date, api_conn.host, api_conn.login, api_conn.extra_dejson['user_email'])
+        return request_id
+    
+    @task(retries=2) 
+    def get_download_link(request_id: str, access_token: str):
+        api_conn = BaseHook.get_connection('here_api_key')
+        download_url = get_download_url(request_id, api_conn.extra_dejson['status_base_url'], access_token, api_conn.login)
+        return download_url
+    
+    access_token = send_request()
+    request_id =  get_request_id(access_token)
+    download_url = get_download_link(request_id, access_token)
+
+    load_data_run = BashOperator(
+        task_id = "load_data",
+        bash_command = '''curl $DOWNLOAD_URL | gunzip | psql -h $HOST -U $USER -d bigdata -c "\COPY here.ta_view FROM STDIN WITH (FORMAT csv, HEADER TRUE);" ''', 
+        env = {"DOWNLOAD_URL": download_url, 
+               "HOST":  BaseHook.get_connection("here_bot").host, 
+               "USER" :  BaseHook.get_connection("here_bot").login, 
+               "PGPASSWORD": BaseHook.get_connection("here_bot").password},
+        append_env=True
+    )
+
+    #Can implement the following when we upgrade to 2.9.1 
+    #@task.bash(env={"DOWNLOAD_URL": download_url,
+    #                "HOST":  BaseHook.get_connection("here_bot").host,
+    #                "USER" :  BaseHook.get_connection("here_bot").login,
+    #           "        PGPASSWORD": BaseHook.get_connection("here_bot").password})
+    #def load_data_run()->str:
+    #    return '''curl $DOWNLOAD_URL | gunzip | psql -h $HOST -U $USER -d bigdata -c "\COPY here.ta_view FROM STDIN WITH (FORMAT csv, HEADER TRUE);" '''
+    # Create a task group for triggering the DAGs
+    @task_group(group_id='trigger_dags_tasks')
+    def trigger_dags(**kwargs):
+        # Define TriggerDagRunOperator for each dag to trigger
+        trigger_operators = []
+        #ds = kwargs['ds']
+        #upstream_dag = kwargs['dag'].dag_id
+        #get dags to trigger from airflow variable 
+        DAGS_TO_TRIGGER = Variable.get('here_dag_triggers', deserialize_json=True)
+        for dag_id in DAGS_TO_TRIGGER:
+            trigger_operator = TriggerDagRunOperator(
+                task_id=f'trigger_{dag_id}',
+                trigger_dag_id=dag_id,
+                execution_date = '{{ ds }}', 
```

Comment:
I'm confused if this is the right date.. let's wait and see tomorrow

Reply:
I reran the yesterday one and its correct! Lets wait till 4:30pm today to check again